File: leetcode/algorithms/111_minimum_depth_of_binary_tree/main.cs
using System;

public class MinimumDepthOfBinaryTree
{
    public class TreeNode
    {
        public int val;
        public TreeNode left;
        public TreeNode right;
        public TreeNode(int val = 0, TreeNode left = null, TreeNode right = null)
        {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public int MinDepth(TreeNode root)
    {
        if (root == null)
        {
            return 0;
        }

        int left = MinDepth(root.left);
        int right = MinDepth(root.right);

        if (left == 0 && right == 0)
        {
            return 1;
        }
        else if (left == 0)
        {
            return right + 1;
        }
        else if (right == 0)
        {
            return left + 1;
        }

        return Math.Min(left, right) + 1;
    }


    // Best Solution
    // Best Solution 1: DFS (Depth-First Search)
    public int bestSolution1(TreeNode root)
    {
        if (root == null)
        {
            return 0;
        }

        var leftMinDepth = MinDepth(root.left);
        var rightMinDepth = MinDepth(root.right);

        if (leftMinDepth == 0 || rightMinDepth == 0)
        {
            return Math.Max(leftMinDepth, rightMinDepth) + 1;
        }

        return Math.Min(leftMinDepth, rightMinDepth) + 1;
    }

    // Best Solution 2: BFS (Breadth-First Search)
    public int bestSolution2(TreeNode root)
    {
        if (root == null)
        {
            return 0;
        }

        var queue = new Queue<TreeNode>();
        queue.Enqueue(root);

        var depth = 0;

        while (queue.Count != 0)
        {
            var size = queue.Count;

            for (var i = 0; i < size; i++)
            {
                var current = queue.Dequeue();

                if (current.left != null)
                {
                    queue.Enqueue(current.left);
                }

                if (current.right != null)
                {
                    queue.Enqueue(current.right);
                }

                if (current.left == null && current.right == null)
                {
                    return depth + 1;
                }
            }

            depth++;
        }

        return 0;
    }
}

File: leetcode/algorithms/111_minimum_depth_of_binary_tree/main.java
import java.util.LinkedList;
import java.util.Queue;

class MinimumDepthOfBinaryTree {
  public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
      this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
      this.val = val;
      this.left = left;
      this.right = right;
    }
  }

  /**
   * Recursion: DFS
   * - Time Complexity: O(N)
   * - Space Complexity: O(H)
   */
  public int minDepth(TreeNode root) {
    if (root == null) {
      return 0;
    }

    int leftDepth = minDepth(root.left);
    int rightDepth = minDepth(root.right);

    if (leftDepth == 0 || rightDepth == 0) {
      return Math.max(leftDepth, rightDepth) + 1;
    }

    return Math.min(leftDepth, rightDepth) + 1;
  }


  // Solution
  /**
   * Solution 1
   * 
   * Iteration: BFS (Queue)
   * - Time Complexity: O(N)
   * - Space Complexity: O(W)
   */
  public int solution1(TreeNode root) {
    if (root == null) {
      return 0;
    }

    Queue<TreeNode> queue = new LinkedList<>();
    queue.offer(root);
    int depth = 1;

    while (!queue.isEmpty()) {
      int levelSize = queue.size();

      for (int i = 0; i < levelSize; i++) {
        TreeNode currentNode = queue.poll();

        if (currentNode.left == null && currentNode.right == null) {
          return depth;
        }

        if (currentNode.left != null) {
          queue.offer(currentNode.left);
        }
        if (currentNode.right != null) {
          queue.offer(currentNode.right);
        }
      }

      depth++;
    }

    return depth;
  }

  /**
   * Solution 2
   * 
   * Recursion: DFS
   * - Time Complexity: O(N)
   * - Space Complexity: O(H)
   */
  public int solution2(TreeNode root) {
    if (root == null) {
      return 0;
    }

    int leftDepth = solution2(root.left);
    int rightDepth = solution2(root.right);

    if (leftDepth == 0 || rightDepth == 0) {
      return leftDepth + rightDepth + 1;
    } else {
      return Math.min(leftDepth, rightDepth) + 1;
    }
  }
}
